compiler/rustc_mir_transform/src/liveness.rs

```diff
@@ -0,0 +1,1087 @@
+                match kind {
+                    AccessKind::Assign | AccessKind::Param => {
+                        err.help("maybe it is overwritten before being read?")
+                    }
+                    AccessKind::Capture => {
+                        err.help("did you mean to capture by reference instead?")
+                    }
+                };
+                err.emit();
+            })
+        }
+    }
+
+    debug!(?dead_captures);
+    dead_captures
+}
+
+/// Small helper to make semantics easier to read.
+#[inline]
+fn is_capture(place: PlaceRef<'_>) -> bool {
+    if !place.projection.is_empty() {
+        debug_assert_eq!(place.local, ty::CAPTURE_STRUCT_LOCAL);
+        true
+    } else {
+        false
+    }
+}
+
+/// Give a diagnostic when any of the string constants look like a naked format string that would
+/// interpolate our dead local.
+fn maybe_suggest_literal_matching_name(
+    err: &mut DiagnosticBuilder<'_, ()>,
+    name: &str,
+    body: &Body<'_>,
+) {
+    struct LiteralFinder {
+        name: String,
+        name_colon: String,
+        found: Option<Location>,
+    }
+
+    impl<'tcx> Visitor<'tcx> for LiteralFinder {
+        fn visit_constant(&mut self, constant: &Constant<'tcx>, loc: Location) {
+            if let ty::Ref(_, ref_ty,_ ) = constant.ty().kind() && ref_ty.kind() == &ty::Str {
+                let rendered_constant = constant.literal.to_string();
+                if rendered_constant.contains(&self.name)
+                    || rendered_constant.contains(&self.name_colon)
+                {
+                    self.found = Some(loc);
+                }
+            }
+        }
+    }
+
+    let mut finder = LiteralFinder {
+        name: format!("{{{name}}}"),
+        name_colon: format!("{{{name}:"),
+        found: None,
+    };
+    finder.visit_body(body);
+    if let Some(loc) = finder.found {
+        let lit_span = body.source_info(loc).span;
+
+        err.span_label(
+            lit_span,
+            "you might have meant to use string interpolation in this string literal",
+        );
+        err.multipart_suggestion(
+            "string interpolation only works in `format!` invocations",
+            vec![
+                (lit_span.shrink_to_lo(), "format!(".to_string()),
+                (lit_span.shrink_to_hi(), ")".to_string()),
+            ],
+            Applicability::MachineApplicable,
+        );
+    }
+}
+
+/// Compute self-assignments of the form `a += b`.
+///
+/// MIR building generates 2 statements and 1 terminator for such assignments:
+/// - _temp = CheckedBinaryOp(a, b)
+/// - assert(!_temp.1)
+/// - a = _temp.0
+///
+/// This function tries to detect this pattern in order to avoid marking statement as a definition
+/// and use.  This will let the analysis be dictated by the next use of `a`.
+///
+/// Note that we will still need to account for the use of `b`.
```

Comment:
To mention potential future directions: the handling of updates like `a += 1` would follow naturally from strong liveness analysis, where in an assignment `a = op(b, c)`, where `op` has no side-effects, variables `b` and `c` are strongly live if `a` was strongly live before.